NeoML/src/Dnn/ReferenceDnnFactory.cpp

```diff
@@ -118,18 +121,32 @@ void CReferenceDnnFactory::serialize( CArchive& archive, bool optimizeDnn )
 
 	NeoAssert( archive.IsLoading() );
 	Origin->Dnn.Serialize( archive );
-	archive.Close();
 
 	if( optimizeDnn == true ) {
+		NeoAssert( !Origin->Dnn.IsReferenceDnn() );
 		( void ) OptimizeDnn( Origin->Dnn );
 	}
 
 	// And revert back the restrictions
+	allowLayersToShareParamBlobs( Origin->Dnn );
 	Origin->Dnn.DisableLearning();
 	swap( tmp, Origin->Dnn.referenceDnnInfo );
 	NeoAssert( Origin->Dnn.IsReferenceDnn() );
 }
 
+void CReferenceDnnFactory::allowLayersToShareParamBlobs( CDnn& dnn )
+{
+	for( CPtr<CBaseLayer>& layer : dnn.layers ) {
+		CMultichannelLookupLayer* lookup = dynamic_cast<CMultichannelLookupLayer*>( layer.Ptr() );
```

Comment:
```suggestion
		auto* lookup = dynamic_cast<CMultichannelLookupLayer*>( layer.Ptr() );
```
(very minor) avoid type duplication in one string

Reply:
fixed